packages/story-editor/src/app/taxonomy/taxonomyProvider.js

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+import {
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from '@web-stories-wp/react';
+/**
+ * Internal dependencies
+ */
+import cleanForSlug from '../../utils/cleanForSlug';
+import { useAPI } from '../api';
+import { useStory } from '../story';
+import Context from './context';
+import {
+  dictonaryOnKey,
+  mapObjectVals,
+  mergeNestedDictionaries,
+  objectFromEntries,
+  mapObjectKeys,
+  cacheFromEmbeddedTerms,
+} from './utils';
+
+function TaxonomyProvider(props) {
+  const [taxonomies, setTaxonomies] = useState([]);
+  const [selectedSlugs, setSelectedSlugs] = useState({});
+  const [termCache, setTermCache] = useState({});
+  const { updateStory, isStoryLoaded, story } = useStory(
+    ({ state: { pages, story }, actions: { updateStory } }) => ({
+      updateStory,
+      isStoryLoaded: pages.length > 0,
+      story,
+    })
+  );
+
+  const { getTaxonomyTerm, createTaxonomyTerm, getTaxonomies } = useAPI(
+    ({ actions }) => actions
+  );
+
+  // Get all registered `web-story` taxonomies.
+  useEffect(() => {
+    (async function () {
+      try {
+        const result = await getTaxonomies();
+        setTaxonomies(Object.values(result));
+      } catch (e) {
+        // Do we wanna do anything here?
+      }
+    })();
+  }, [getTaxonomies]);
+
+  // Reference embedded terms in the story and taxonomies
+  // to get the initial selected terms as well as populate
+  // the taxonomy term cache
+  const hasHydratedRunOnce = useRef(false);
```

Comment:
💭 nit: this feels a little more grammatically correct to me, but I don't think it matters
```suggestion
  const hasHydrationRunOnce = useRef(false);
```